### PR TITLE
refactor(dialects/abstract): asyncify methods

### PR DESCRIPTION
**Pull request by jedwards1211**
_Please make sure to review and check all of these items:_

- [ ] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

File: lib/dialects/abstract/connection-manager.js [hunk 1]
```diff
@@ -91,29 +91,31 @@ class ConnectionManager {
...
    * @returns {Promise}
    */
-  close() {
+  async close() {
     // Mark close of pool
-    this.getConnection = function getConnection() {
-      return Promise.reject(new Error('ConnectionManager.getConnection was called after the connection manager was closed!'));
+    this.getConnection = async function getConnection() {
+      throw new Error(
+        new Error('ConnectionManager.getConnection was called after the connection manager was closed!')
```
Comment (sushantdhiman) [inline lib/dialects/abstract/connection-manager.js:114]: Double `new Error` statement, remove one please

Comment (jedwards1211) [reply lib/dialects/abstract/connection-manager.js:114]: this was a bug in my automated conversion of `return Promise.reject`, tomorrow I'll double-check my other PRs to make sure no instances of this mistake escaped testing

**Comment by SimonSchick** [comment]
I wonder if this is wise, before doing this we should probably drop bluebird as the promise implementation to avoid mixing bluebird and native promises.

I did some work on this on another branch but never got around to finish it.

**Comment by jedwards1211** [comment]
I've tracked down and replaced all of those cases I could find so far, though I guess it's true that it would be good to remove Bluebird entirely

**Comment by SimonSchick** [comment]
@jedwards1211 if you are bored https://github.com/sequelize/sequelize/tree/chore/remove-bluebird

**Comment by jedwards1211** [comment]
@SimonSchick @sushantdhiman 
The only remaining Bluebird API usage in `lib` on master that I found is:
```
lib/dialects/mssql/query.js Position { line: 117, column: 17, token: 714 } Promise.using
lib/sequelize.js Position { line: 1144, column: 4, token: 3317 } Promise.config
```
do you know what exactly the purpose of the `Promise.using` stuff for MSSQL is?  (`lib/dialects/mssql/resource-lock.js`)
I'm a bit confused what that code is doing.

The remaining Bluebird API usage in `test` is a bunch of `join()` calls, one case of `Promise.onPossiblyUnhandledRejection`, and a few `Promise.using` calls for MSSQL.

**Comment by SimonSchick** [comment]
`.using` is similar to `using` in c#, it will call disposer function when exiting the scope, which is meant to 'cleanup' resources eg. close connections/file handles etc.
See https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.8 and
http://bluebirdjs.com/docs/api/promise.using.html

**Comment by jedwards1211** [comment]
@SimonSchick I sort of get that but is it just an overcomplicated method of ensuring two queries aren't run at once on the same connection?  It doesn't actually seem like it's being used to dispose of resources.  The `disposer` is just used to resolve the previous promise:

```js
  lock() {
    const lock = this.previous;
    let resolve;
    this.previous = new Promise(r => {
      resolve = r;
    });
    return lock.disposer(resolve);
  }
```

**Comment by jedwards1211** [comment]
Seems like kind of a mess because this `ResourceLock` object which isn't actually a connection but a wrapper gets passed to `this.pool.destroy`, and goes on to `ConnectionManager._disconnect`, which passes it to `beforeDisconnect` and `afterDisconnect` hooks.

**Comment by jedwards1211** [comment]
Seems to me like the cleanest replacement would be to wrap the `connection` using `Object.create` to add a `waitUntilIdle` method.

**Comment by jedwards1211** [comment]
We could use [`p-limit`](https://github.com/sindresorhus/p-limit) to instead to prevent more than one query from running at once on a connection

**Comment by jedwards1211** [comment]
all of these PRs should be ready to go

**Comment by sushantdhiman** [comment]
@jedwards1211 / @SimonSchick we discussed a bit here https://github.com/sequelize/sequelize/issues/11859

Ideally we can remove Bluebird first and then convert to async methods, but for some cases intermediate code may become messy. So I think removing Bluebird API usage from core code and async conversion is faster.

We will remove bluebird after everything is converted, so there will be no promise mixing

**Comment by SimonSchick** [comment]
@sushantdhiman should I bring my bluebird removal branch up to date?
Once that is done we can utility automated tools to perform async/await conversion in many places!

**Comment by sushantdhiman** [comment]
@SimonSchick that branch contains other changes as well, also too big to review. It is better to handle this in small PRs.

**Comment by papb** [comment]
Hi @SimonSchick I forgot you had that branch. Please see with @jedwards1211 if it is feasible to use part of your work in the work he is currently doing.

**Comment by SimonSchick** [comment]
A lot of that is probably still usable.
The branch is kind of a mush (got very carried away).
You can definitely pick and choose individual files and do a quick sanity check and them use them, I wouldn't blindly copy things however.

The branch also still contains stuff like the hooks refactor (which I'd love to land in v6 as it DOES do a lot of streamlining).

**Comment by papb** [comment]
> The branch also still contains stuff like the hooks refactor (which I'd love to land in v6 as it DOES do a lot of streamlining).

Awesome, I agree the sequelize codebase desperately needs lots of refactoring, so that's great. Let's wait for @jedwards1211 to finish his super work on async/await and then we can look at these other refactors

**Comment by jedwards1211** [comment]
@SimonSchick it's not worth it I'm afraid, as a lot of PRs I've made that have been merged have already eliminated nearly all of the bluebird usage.

**Comment by jedwards1211** [comment]
Oh I didn't see what you said about hooks refactoring though.  How heavily does it depend on the changes to remove bluebird in your branch?

**Pull request merged**